When building with the option `char_type=wchar_t`, only the define will not be enough, both Linux and Mac (non-Windows), will fail with:

```
conan create all/conanfile.py --version=2.8.1 -o "&:char_type=wchar_t"
 

======== Exporting recipe to the cache ========
expat/2.8.1: Exporting package recipe: /home/conan/project/all/conanfile.py
expat/2.8.1: exports: File 'conandata.yml' found. Exporting it...
expat/2.8.1: Copied 1 '.py' file: conanfile.py
expat/2.8.1: Copied 1 '.yml' file: conandata.yml
expat/2.8.1: Exported to cache folder: /root/.conan2/p/expat89e39aef6006c/e
expat/2.8.1: Exported: expat/2.8.1#6e229e21860c73dc7bf0118f776b0ca6 (2026-06-03 11:51:09 UTC)

======== Input profiles ========
Profile host:
[settings]
arch=armv8
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=13
os=Linux
[options]
&:char_type=wchar_t

Profile build:
[settings]
arch=armv8
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=13
os=Linux


======== Computing dependency graph ========
Graph root
    cli
Requirements
    expat/2.8.1#6e229e21860c73dc7bf0118f776b0ca6 - Cache
Build requirements
    cmake/4.3.2#5bae0b181463ca4dc3e924a981367ee6 - Cache
Resolved version ranges
    cmake/[>=3.17]: cmake/4.3.2

======== Computing necessary packages ========
expat/2.8.1: Forced build from source
Requirements
    expat/2.8.1#6e229e21860c73dc7bf0118f776b0ca6:5a4b5c2361d4440f744e923ed37543e5c9eccd37 - Build
Build requirements
    cmake/4.3.2#5bae0b181463ca4dc3e924a981367ee6:c93719558cf197f1df5a7f1d071093e26f0e44a0#85ca38a2fffde6c18c8b0fc26b300cdc - Cache

======== Installing packages ========
cmake/4.3.2: Already installed! (1 of 2)

-------- Installing package expat/2.8.1 (2 of 2) --------
expat/2.8.1: Building from source
expat/2.8.1: Package expat/2.8.1:5a4b5c2361d4440f744e923ed37543e5c9eccd37
expat/2.8.1: Copying sources to build folder
expat/2.8.1: Building your package in /root/.conan2/p/b/expat81cc1a54bff90/b
expat/2.8.1: Calling generate()
expat/2.8.1: Generators folder: /root/.conan2/p/b/expat81cc1a54bff90/b/build/Release/generators
expat/2.8.1: CMakeToolchain generated: conan_toolchain.cmake
expat/2.8.1: CMakeToolchain generated: /root/.conan2/p/b/expat81cc1a54bff90/b/build/Release/generators/CMakePresets.json
expat/2.8.1: CMakeToolchain generated: /root/.conan2/p/b/expat81cc1a54bff90/b/src/CMakeUserPresets.json
expat/2.8.1: Generating aggregated env files
expat/2.8.1: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
expat/2.8.1: Calling build()
expat/2.8.1: Running CMake.configure()
expat/2.8.1: RUN: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/root/.conan2/p/b/expat81cc1a54bff90/p" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/root/.conan2/p/b/expat81cc1a54bff90/b/src"
-- Using Conan toolchain: /root/.conan2/p/b/expat81cc1a54bff90/b/build/Release/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The C compiler identification is GNU 13.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at CMakeLists.txt:255 (message):
  Configuration -DEXPAT_CHAR_TYPE=wchar_t requires
  -DCMAKE_{C,CXX}_FLAGS=-fshort-wchar (which was not found) and libc compiled
  with -fshort-wchar, too.
```

The check comes from https://github.com/libexpat/libexpat/blob/R_2_8_1/expat/CMakeLists.txt#L255

```cmake
elseif(EXPAT_CHAR_TYPE STREQUAL "wchar_t")
    set(_EXPAT_UNICODE ON)
    set(_EXPAT_UNICODE_WCHAR_T ON)
    if(NOT WIN32)
        string(FIND "${CMAKE_C_FLAGS}" "-fshort-wchar" _expat_short_wchar_found)
        if(${_expat_short_wchar_found} EQUAL "-1")
            message(SEND_ERROR "Configuration -DEXPAT_CHAR_TYPE=wchar_t requires -DCMAKE_{C,CXX}_FLAGS=-fshort-wchar (which was not found) and libc compiled with -fshort-wchar, too.")
        endif()
        if (EXPAT_BUILD_TOOLS)
            message(SEND_ERROR "The xmlwf tool can not be built with option -DEXPAT_CHAR_TYPE=wchar_t outside of Windows. Please pass -DEXPAT_CHAR_TYPE=char or -DEXPAT_BUILD_TOOLS=OFF.")
        endif()
    endif()
else()
```

So we need to inject `-fshort-wchar` as well for CFLAGS and forward it. In case of mismatch, the test package/consumer, will fail as well:

```
xpat/2.8.1 (test package): RUN: cmake --build "/Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/build/apple-clang-21-armv8-20-release" --verbose -- -j10
Change Dir: '/Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/build/apple-clang-21-armv8-20-release'

Run Build Command(s): /opt/homebrew/bin/ninja -v -j10
[1/2] /usr/bin/cc -DXML_STATIC -DXML_UNICODE -DXML_UNICODE_WCHAR_T -isystem /Users/uilian/.conan2/p/b/expat980d0b06227c6/p/include -O3 -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk -mmacosx-version-min=13.0 -MD -MT CMakeFiles/test_package.dir/test_package.c.o -MF CMakeFiles/test_package.dir/test_package.c.o.d -o CMakeFiles/test_package.dir/test_package.c.o -c /Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/test_package.c
FAILED: [code=1] CMakeFiles/test_package.dir/test_package.c.o 
/usr/bin/cc -DXML_STATIC -DXML_UNICODE -DXML_UNICODE_WCHAR_T -isystem /Users/uilian/.conan2/p/b/expat980d0b06227c6/p/include -O3 -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk -mmacosx-version-min=13.0 -MD -MT CMakeFiles/test_package.dir/test_package.c.o -MF CMakeFiles/test_package.dir/test_package.c.o.d -o CMakeFiles/test_package.dir/test_package.c.o -c /Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/test_package.c
In file included from /Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/test_package.c:1:
In file included from /Users/uilian/.conan2/p/b/expat980d0b06227c6/p/include/expat.h:50:
/Users/uilian/.conan2/p/b/expat980d0b06227c6/p/include/expat_external.h:137:8: error: "sizeof(wchar_t) != 2; Need -fshort-wchar for both Expat and libc"
  137 | #      error "sizeof(wchar_t) != 2; Need -fshort-wchar for both Expat and libc"
      |        ^
1 error generated.
ninja: build stopped: subcommand failed.
```

----

Later, to validate the issue #30226, I modified the test_package.c and built the recipe without export `XML_UNICODE_WCHAR_T` definition:

```c
#include "expat.h"

static const XML_Char document[] = L"<root/>";
static const XML_LChar label[] = L"ok";

int
main(void)
{
  XML_Parser parser = XML_ParserCreate(NULL);
  const int len = (int)(sizeof(document) - sizeof(XML_Char));

  (void)label;
  if (XML_Parse(parser, (const char *)document, len, 1) != XML_STATUS_OK)
    return 1;
  XML_ParserFree(parser);
  return 0;
}
```

Then, the test package failed:

```
expat/2.8.1 (test package): RUN: cmake --build "/Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/build/apple-clang-21-armv8-20-release" --verbose -- -j10
Change Dir: '/Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/build/apple-clang-21-armv8-20-release'

Run Build Command(s): /opt/homebrew/bin/ninja -v -j10
[1/2] /usr/bin/cc -DXML_STATIC -DXML_UNICODE -isystem /Users/uilian/.conan2/p/b/expatdaeff139d9b92/p/include -O3 -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk -mmacosx-version-min=13.0 -fshort-wchar -MD -MT CMakeFiles/test_package.dir/test_package.c.o -MF CMakeFiles/test_package.dir/test_package.c.o.d -o CMakeFiles/test_package.dir/test_package.c.o -c /Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/test_package.c
FAILED: [code=1] CMakeFiles/test_package.dir/test_package.c.o 
/usr/bin/cc -DXML_STATIC -DXML_UNICODE -isystem /Users/uilian/.conan2/p/b/expatdaeff139d9b92/p/include -O3 -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk -mmacosx-version-min=13.0 -fshort-wchar -MD -MT CMakeFiles/test_package.dir/test_package.c.o -MF CMakeFiles/test_package.dir/test_package.c.o.d -o CMakeFiles/test_package.dir/test_package.c.o -c /Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/test_package.c
/Users/uilian/Development/conan/conan-center-index/recipes/expat/all/test_package/test_package.c:4:24: error: initializing char array with wide string literal
    4 | static const XML_LChar label[] = L"ok";
      |                        ^
1 error generated.
ninja: build stopped: subcommand failed.
```

There is a mismatch here, it treats XML_LChar as char while the built library was built for wchar_t text size. When using the patch from this PR, it works good, full build log: [expat-2.8.1-mac-armv8-clang21-release-static-wchart.log](https://github.com/user-attachments/files/28551798/expat-2.8.1-mac-armv8-clang21-release-static-wchart.log)

--- a/recipes/expat/all/conanfile.py
+++ b/recipes/expat/all/conanfile.py
@@ -62,8 +62,6 @@ class ExpatConan(ConanFile):
             tc.variables["EXPAT_MSVC_STATIC_CRT"] = is_msvc_static_runtime(self)
         tc.variables["EXPAT_BUILD_PKGCONFIG"] = False
         tc.variables["EXPAT_LARGE_SIZE"] = self.options.large_size
-        if self.options.char_type == "wchar_t" and self.settings.os != "Windows":
-            tc.extra_cflags.append("-fshort-wchar")
         tc.generate()
 
     def build(self):
@@ -94,9 +92,6 @@ class ExpatConan(ConanFile):
             self.cpp_info.defines.append("XML_UNICODE")
         if self.options.get_safe("char_type") == "wchar_t":
             self.cpp_info.defines.append("XML_UNICODE_WCHAR_T")
-            if self.settings.os != "Windows":
-                self.cpp_info.cflags.append("-fshort-wchar")
-                self.cpp_info.cxxflags.append("-fshort-wchar")
         if self.options.large_size:
             self.cpp_info.defines.append("XML_LARGE_SIZE")
 

